### PR TITLE
First draft

### PR DESCRIPTION
**Pull request by sblack4**

### CODE REVIEW

**Review by cdaniluk — COMMENTED** [review]
I'm having a philosophical problem with the name field. Specifically, many modules that we pass tags to will need to do their own thing with the Name field. Some name fields are constrained in length and sometimes we can't set a name at all.

I suppose those modules could overload the map with their own name and/or remove name outright when needed.

File: variables.tf [hunk 1]
```diff
@@ -0,0 +1,30 @@
+
+variable "tags" {
+  type        = map(string)
+  default     = {}
+  description = "Additional tags. Overwrite module-generated tags"
```
Comment (cdaniluk) [inline variables.tf:5]: It looks like this will overwrite values for existing keys. But it won't squash unwanted keys.

**Comment by sblack4** [comment]
Yeah - we can't anticipate every need, only do our best to help those who are actually using our module. 
My main concern was over what to enforce as a tag. e.g. `env`, `namespace`, `owner`, etc. I'd rather not enforce anything but then I'm not sure this module has much of a point.

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,31 @@
+
+locals {
+  enforce_uppercase = lower(var.enforce_case) == "upper"
+  enforce_lowercase = lower(var.enforce_case) == "lower"
+
+  uncased_name = join(var.name_seperator, var.names)
+  name         = local.enforce_uppercase ? upper(local.uncased_name) : local.enforce_lowercase ? lower(local.uncased_name) : local.uncased_name
```
Comment (smiller171) [inline main.tf]:
```suggestion
  name = (
    local.enforce_uppercase ?
      upper(local.uncased_name) :
      local.enforce_lowercase ?
        lower(local.uncased_name) :
        local.uncased_name
  )
```

File: variables.tf [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+variable "enforce_case" {
+  type        = string
+  default     = "None"
+  description = "Type casing to enforce. Options are: [ None (default) | UPPER | lower ]"
+}
+
+variable "names" {
+  type        = set(string)
```
Comment (smiller171) [inline variables.tf]: This should be a list, not a set. Maintaining the order is more important than making sure they are unique (sets are unordered)
```suggestion
  type        = list(string)
```

File: main.tf (hunk 2 above)
Comment (smiller171) [inline main.tf]: let's break up this long line

```suggestion
  name = (
    local.enforce_uppercase ?
      upper(local.uncased_name) :
      local.enforce_lowercase ?
        lower(local.uncased_name) :
        local.uncased_name
  )
```

**Review by smiller171 — CHANGES_REQUESTED** [review]

**Comment by smiller171** [comment]
> I'm having a philosophical problem with the name field. Specifically, many modules that we pass tags to will need to do their own thing with the Name field. Some name fields are constrained in length and sometimes we can't set a name at all.
> 
> I suppose those modules could overload the map with their own name and/or remove name outright when needed.

I think the length restriction (especially when using `prefix`) is the most important thing here. If we reverse the order of name then it's easy to aplly `substr` to get a value of the length needed while still maintaining as much meaning as possible (you want more specificity if you have to choose I think)

**Review by smiller171 — APPROVED** [review]
LGTM

File: main.tf (hunk 2 above)
Comment (sblack4) [reply main.tf]: This is more readable. But isn't code that's hard to write supposed to be hard to read?

File: variables.tf [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  type        = map(string)
+  default     = {}
+  description = "Additional tags. Overwrite module-generated tags"
+}
+
+variable "timestamp_format" {
+  type        = string
+  default     = "YYYY-MM-DD'T'hh:mm:ssZ"
+  description = "Timestamp format, see terraform.io/docs/configuration/functions/formatdate.html#specification-syntax"
+}
```
Comment (sblack4) [inline variables.tf]:
```suggestion

```
I really like having gratuitous timestamps but we can get rid of this now.

File: examples/basic/README.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      "key" = "terraform_root_module"
+      "value" = "."
+    },
+    {
+      "key" = "terraform_workspace"
+      "value" = "default"
+    },
+  ]
+}
+```
```
Comment (sblack4) [inline examples/basic/README.md:45]: let's def try to keep the example updated, it'd be nice if we had this done automatically on merge or something

**Review by sblack4 — COMMENTED** [review]
:+1:

**Review by cdaniluk — APPROVED** [review]

**Pull request merged**